### PR TITLE
Implement Proper Record Copying Behaviour for ROCK

### PR DESCRIPTION
**Pull request by bska**
This PR adds the expected behaviour for all-defaulted records in `ROCK`, provided the all-defaulted records are not the first of the keyword.  Similarly to, e.g. `PVTW`, all-defaulted records are treated as copies of the immediately preceding record.

In other words, given
```
ROCK
-- REF. PRES   COMPRESSIBILITY
  280.000        5.6E-5 /
/
```
the second record is supposed to be a copy of the first.

### CODE REVIEW

**Comment by bska** [comment]
The effect of not properly implementing the copying behaviour can be quite dramatic, as can be seen from the example summary curves below using OPM-Test's [model1/ROCKOPTS_MODEL_1](https://github.com/OPM/opm-tests/blob/93d63368497abd5c4be3491719171edbeda81e73/model1/ROCKOPTS_MODEL_1.DATA) simulation model (PR OPM/opm-tests#889).  The `_MASTER` and `_DEV` runs were conducted using the current master and this PR's development sources, respectively.  The `_DUPL` run used the current master sources, but amended the input deck to explicitly duplicate/copy the first rock compressibility record into the second.  In other words,
```diff
--- ./ROCKOPTS_MODEL_1_MASTER.DATA	2023-04-20 13:05:28.216204991 +0200
+++ ./ROCKOPTS_MODEL_1_DUPL.DATA	2023-04-20 13:15:09.786450571 +0200
@@ -210,7 +210,7 @@
 ROCK
 -- REF. PRES   COMPRESSIBILITY
    280.000        5.6E-5 /
-/  -- default, copy of previous record
+   280.000        5.6E-5 /
```

I have added time-step markers to the `_DUPL` curves to suggest that the `_DEV` curves coincide with those of the `_DUPL` run.

This PR is nevertheless a work in progress, both because it depends on and contains #3479 and therefore must not be merged before that PR, but also because it will benefit from additional testing. I will keep the PR in "draft" state until such time as it is ready for review.

---

### Field Level Quantities ###

Field

### Well Level Quantities ###

PROD_2

**Comment by tskille** [comment]
This PR has now been tested on the full field model where the problem was identified and results are excellent. I'm looking forward to see this PR go into master.

**Comment by bska** [comment]
> This PR has now been tested on the full field model where the problem was identified and results are excellent

Thanks a lot for testing the PR, that is greatly appreciated.

> I'm looking forward to see this PR go into master.

With those positive testing results from the full field model in hand I'll prioritise getting the work into the master branch now.

**Comment by bska** [comment]
The upstream PR has been merged.  I'm marking this as "ready for review" now.

**Comment by bska** [comment]
jenkins build this please

**Review by blattms — APPROVED** [review]
Just one question for understanding. This is good as it is.

**Comment by bska** [comment]
jenkins build this please

File: opm/input/eclipse/EclipseState/Tables/FlatTable.hpp [hunk 1]
```diff
@@ -313,13 +313,22 @@ struct ROCKRecord {
...
     static RockTable serializationTestObject()
     {
         return RockTable({{1.0, 2.0}});
     }
+
+    template <class Serializer>
+    void serializeOp(Serializer& serializer)
+    {
+        FlatTableWithCopy::serializeOp(serializer);
+    }
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/input/eclipse/EclipseState/Tables/FlatTable.hpp]: > Wouldn't this be the default behavior if the method would just be there in `FieldTableWithCopy`

You're absolutely right; good catch!  I've pushed an update to remove the unnecessary implementations of `serializeOp()` and the latest [build check](https://ci.opm-project.org/job/opm-common-PR-builder/5693/) shows that these functions indeed were unneeded.

[the remark this answers is not in the document]
Comment (blattms) [reply opm/input/eclipse/EclipseState/Tables/FlatTable.hpp]: thanks. Was a bit worried that I might make a fool out of myself in public again :sweat_smile: In it goes.

**Pull request merged**